File: nim/src/algorithms/smith_waterman.nim
from std/unicode import reversed

type
  ScoreInfo* = object
    match_bonus*: int
    mismatch_penalty*: int
    gap_penalty*: int

func create_matrix(subject_length: int, query_length: int): seq[int] =
  return newSeq[int]((subject_length + 1) * (query_length + 1))

func idx(columnWidth: int, i: int, j: int): int =
  return i * columnWidth + j

func matrix_get(matrix: seq[int], column_width, i, j: auto): auto =
  return matrix[idx(column_width, i, j)]

func matrix_set(matrix: var seq[int], column_width, i, j: auto, value: int) =
  matrix[idx(column_width, i, j)] = value

func calculate_matrix(
    subject, query: string,
    matrix: var seq[int],
    score_info: ScoreInfo,
) =
  let subject_length = subject.len()
  let query_length = query.len()
  let column_width = query_length + 1

  for i in countup(1, subject_length):
    for j in countup(1, query_length):
      let diagonal = matrix_get(matrix, column_width, i - 1, j - 1)
      let up = matrix_get(matrix, column_width, i - 1, j)
      let left = matrix_get(matrix, column_width, i, j - 1)

      let matching =
        if subject[i - 1] == query[j - 1]:
          diagonal + score_info.match_bonus
        else:
          diagonal - score_info.mismatch_penalty

      let upward_gap = up - score_info.gap_penalty
      let leftward_gap = left - score_info.gap_penalty

      matrix_set(matrix, column_width, i, j, [matching, upward_gap, leftward_gap, 0].max())

proc backtrack_alignment(
    subject, query: string,
    matrix: seq[int],
    score_info: ScoreInfo,
): auto =
  let subject_length = subject.len()
  let query_length = query.len()
  let column_width = query_length + 1

  let max_score_idx = matrix.find(matrix.max())

  var subject_alignment = ""
  var query_alignment = ""

  var i = max_score_idx div column_width
  var j = max_score_idx mod column_width

  while i > 0 and j > 0:
    let current = matrix_get(matrix, column_width, i, j)

    let diagonal = matrix_get(matrix, column_width, i - 1, j - 1)
    let diagonal_match = diagonal + score_info.match_bonus
    let diagonal_mismatch = diagonal - score_info.mismatch_penalty


    let up = matrix_get(matrix, column_width, i - 1, j)
    let up_gap = up - score_info.gap_penalty

    let left = matrix_get(matrix, column_width, i, j - 1)
    let left_gap = left - score_info.gap_penalty

    if current == 0:
      break
    elif (current == diagonal_match and subject[i - 1] == query[j - 1]) or
        (current == diagonal_mismatch and subject[i - 1] != query[j - 1]):
      add(subject_alignment, subject[i - 1])
      add(query_alignment, query[j - 1])
      i = i - 1
      j = j - 1
    elif current == up_gap:
      add(subject_alignment, subject[j - 1])
      add(query_alignment, "-")
      i = i - 1
    elif current == left_gap:
      add(subject_alignment, "-")
      add(query_alignment, query[i - 1])
      j = j - 1
    else:
      break

  return (reversed(subject_alignment), reversed(query_alignment))

proc smith_waterman*(subject, query: string, score_info: ScoreInfo): auto =
  var matrix = create_matrix(len(subject), len(query))

  calculate_matrix(subject, query, matrix, score_info)
  return backtrack_alignment(subject, query, matrix, score_info)

File: go/algorithm/smithwaterman.go
package algorithm

import (
	"slices"
	"strings"
)

type ScoreInfo struct {
	MatchBonus      int
	MismatchPenalty int
	GapPenalty      int
}

// this is basically for performance, so wo don't do array
// of arrays
func createMatrix(subjectLength int, queryLength int) []int {
	return make([]int, (subjectLength+1)*(queryLength+1))
}

func getMatrixAccessors(matrix []int, queryLength int) (func(int, int) int, func(int, int, int)) {
	get := func(i int, j int) int {
		return matrix[i*(queryLength+1)+j]
	}
	set := func(i int, j int, value int) {
		matrix[i*(queryLength+1)+j] = value
	}
	return get, set
}

func calculateMatrix(matrix []int, subject string, query string, scoreInfo ScoreInfo) {
	get, set := getMatrixAccessors(matrix, len(query))

	for i := 1; i <= len(subject); i++ {
		for j := 1; j <= len(query); j++ {
			diagonal := get(i-1, j-1)
			up := get(i-1, j)
			left := get(i, j-1)

			diagonal_move := diagonal
			if subject[i-1] == query[j-1] {
				diagonal_move += scoreInfo.MatchBonus
			} else {
				diagonal_move -= scoreInfo.MismatchPenalty
			}

			up_move := up - scoreInfo.GapPenalty
			left_move := left - scoreInfo.GapPenalty

			set(i, j, max(diagonal_move, up_move, left_move, 0))
		}
	}
}

func reverseString(s string) string {
	result := strings.Builder{}
	runes := []rune(s)
	for _, runeValue := range slices.Backward(runes) {
		result.WriteRune(runeValue)
	}

	return result.String()
}

func backtrackAlignment(matrix []int, subject string, query string, scoreInfo ScoreInfo) (string, string) {
	get, _ := getMatrixAccessors(matrix, len(query))

	maxScoreIndex := 0
	maxScore := 0

	for i := len(matrix) - 1; i >= 0; i-- {
		if matrix[i] > maxScore {
			maxScore = matrix[i]
			maxScoreIndex = i
		}
	}

	subjectBuilder := strings.Builder{}
	queryBuilder := strings.Builder{}

	row := maxScoreIndex / (len(query) + 1)
	col := maxScoreIndex % (len(query) + 1)

	for row > 0 && col > 0 {
		currentScore := get(row, col)
		diagonalScore := get(row-1, col-1)
		upScore := get(row-1, col)
		leftScore := get(row, col-1)

		if currentScore == 0 {
			break
		}

		if currentScore == diagonalScore+scoreInfo.MatchBonus && subject[row-1] == query[col-1] {
			subjectBuilder.WriteByte(subject[row-1])
			queryBuilder.WriteByte(query[col-1])
			row--
			col--
		} else if currentScore == diagonalScore-scoreInfo.MismatchPenalty && subject[row-1] != query[col-1] {
			subjectBuilder.WriteByte(subject[row-1])
			queryBuilder.WriteByte(query[col-1])
			row--
			col--
		} else if currentScore == upScore-scoreInfo.GapPenalty {
			subjectBuilder.WriteByte(subject[row-1])
			queryBuilder.WriteByte('-')
			row--
		} else if currentScore == leftScore-scoreInfo.GapPenalty {
			subjectBuilder.WriteByte('-')
			queryBuilder.WriteByte(query[col-1])
			col--
		} else {
			break
		}
	}

	subjectLine := subjectBuilder.String()
	queryLine := queryBuilder.String()

	return reverseString(subjectLine), reverseString(queryLine)
}

func SmithWaterman(subject string, query string, scoreInfo ScoreInfo) (string, string) {
	matrix := createMatrix(len(subject), len(query))
	calculateMatrix(matrix, subject, query, scoreInfo)
	return backtrackAlignment(matrix, subject, query, scoreInfo)
}
